tools/ja-rule/USBDeviceManager.cpp

```diff
@@ -0,0 +1,259 @@
+using std::auto_ptr;
+
+static const uint16_t kProductId = 0x0053;
+static const uint16_t kVendorId = 0x04d8;
+
+namespace {
+
+int hotplug_callback(OLA_UNUSED struct libusb_context *ctx,
+                     struct libusb_device *dev,
+                     libusb_hotplug_event event,
+                     void *user_data) {
+  USBDeviceManager *manager = reinterpret_cast<USBDeviceManager*>(user_data);
+  manager->HotPlugEvent(dev, event);
+  return 0;
+}
+}  // namespace
+
+USBDeviceManager::USBDeviceManager(SelectServer* ss,
+                                   NotificationCallback* notification_cb)
+    : m_ss(ss),
+      m_cleanup_thread(Thread::Options("cleanup-thread")),
+      m_context(NULL),
+      m_notification_cb(notification_cb),
+      m_suppress_hotplug_events(false) {
+}
+
+USBDeviceManager::~USBDeviceManager() {
+  if (m_context) {
+    Stop();
+  }
+}
+
+bool USBDeviceManager::Start() {
+  int r = libusb_init(&m_context);
+
+  if (r < 0) {
+    OLA_WARN << "libusb_init() failed: " << libusb_error_name(r);
+    m_cleanup_thread.Stop();
+    return false;
+  }
+
+  libusb_set_debug(m_context, 3);
+
+  m_start_thread_id = Thread::Self();
+  m_usb_thread.reset(
+      new LibUsbHotplugThread(m_context, hotplug_callback, this));
+  bool ok = m_usb_thread->Init();
+  if (ok) {
+    m_cleanup_thread.Start();
+  } else {
+    m_usb_thread.reset();
+    libusb_exit(m_context);
+    m_context = NULL;
+  }
+
+  return ok;
+}
+
+bool USBDeviceManager::Stop() {
+  // At this point there may be:
+  //  - notifications queued on the ss.
+  //  - IP in the critical section of HotPlugEvent
+  //  - IP prior to the critical section of HotPlugEvent
+  {
+    MutexLocker locker(&m_mutex);
+    m_suppress_hotplug_events = true;
+  }
+
+  // Now we're left with:
+  //  - Notifications queued on the ss.
+  //  - IP prior to the critical section of HotPlugEvent, these are now
+  //    surpressed.
+  // Drain callbacks on the SS so all notifications complete.
+  m_ss->DrainCallbacks();
+
+  {
+    MutexLocker locker(&m_mutex);
+    DeviceMap::iterator iter = m_devices.begin();
+    for (; iter != m_devices.end(); ++iter) {
+      if (iter->second) {
+        m_notification_cb->Run(DEVICE_REMOVED, iter->second);
+        DeleteDevice(iter->second);
+      }
+    }
+    m_devices.clear();
+  }
+
+  // Blocks until all devices have been deleted.
+  m_cleanup_thread.Stop();
+
+  m_usb_thread->Shutdown();
+  m_usb_thread.reset();
+  libusb_exit(m_context);
+  m_context = NULL;
+  return true;
+}
+
+void USBDeviceManager::HotPlugEvent(struct libusb_device *usb_device,
+                                    libusb_hotplug_event event) {
+  // See the comments under libusb_hotplug_register_callback in the libusb
+  // documentation. There are a number of caveats.
+  struct libusb_device_descriptor descriptor;
+  libusb_get_device_descriptor(usb_device, &descriptor);
+
+  if (descriptor.idVendor != kVendorId || descriptor.idProduct != kProductId) {
+    return;
```

Comment:
I think we should OLA_DEBUG the detected vendor and product, like we do with the USB plugin.